server/v2/commands.go

```diff
@@ -11,62 +11,166 @@ import (
 	"syscall"
 
 	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
 	"github.com/spf13/viper"
 
+	"cosmossdk.io/core/server"
 	"cosmossdk.io/core/transaction"
 	"cosmossdk.io/log"
 )
 
-// Execute executes the root command of an application.
-// It handles adding core CLI flags, specifically the logging flags.
-func Execute(rootCmd *cobra.Command, envPrefix, defaultHome string) error {
-	rootCmd.PersistentFlags().String(FlagLogLevel, "info", "The logging level (trace|debug|info|warn|error|fatal|panic|disabled or '*:<level>,<key>:<level>')")
-	rootCmd.PersistentFlags().String(FlagLogFormat, "plain", "The logging format (json|plain)")
-	rootCmd.PersistentFlags().Bool(FlagLogNoColor, false, "Disable colored logs")
-	rootCmd.PersistentFlags().StringP(FlagHome, "", defaultHome, "directory for config and data")
+type RootCmdBuilder struct {
+	DefaultHome string
+	EnvPrefix   string
 
+	use     string
+	fixture CommandFixture
+	vipr    *viper.Viper
+	logger  log.Logger
+}
+
+type CommandFixture interface {
+	Bootstrap(cmd *cobra.Command) (WritesConfig, error)
+	RootCommand(
+		rootCommand *cobra.Command,
+		subCommand *cobra.Command,
+		logger log.Logger,
+		config server.ConfigMap,
+	) (*cobra.Command, error)
+}
+
+func NewRootCmdBuilder(
+	fixture CommandFixture,
+	use string,
+	defaultHomeBasename string,
+) (*RootCmdBuilder, error) {
+	f := &RootCmdBuilder{
+		fixture: fixture,
+		use:     use,
+	}
+	var err error
+	f.DefaultHome, err = f.DefaultHomeDir(defaultHomeBasename)
+	if err != nil {
+		return nil, err
+	}
+	return f, nil
+}
+
+func (b *RootCmdBuilder) DefaultHomeDir(name string) (string, error) {
+	// get the home directory from the environment variable
+	// to not clash with the $HOME system variable, when no prefix is set
+	// we check the NODE_HOME environment variable
+	homeDir, envHome := "", "HOME"
+	if len(b.EnvPrefix) > 0 {
+		homeDir = os.Getenv(b.EnvPrefix + "_" + envHome)
+	} else {
+		homeDir = os.Getenv("NODE_" + envHome)
+	}
+	if homeDir != "" {
+		return filepath.Clean(homeDir), nil
+	}
+
+	// get user home directory
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(userHomeDir, name), nil
+}
+
+func (b *RootCmdBuilder) Command() (*cobra.Command, error) {
+	cmd := &cobra.Command{Use: b.use, SilenceErrors: true}
+	SetPersistentFlags(cmd.PersistentFlags(), b.DefaultHome)
 	// update the global viper with the root command's configuration
-	viper.SetEnvPrefix(envPrefix)
+	viper.SetEnvPrefix(b.EnvPrefix)
 	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
 	viper.AutomaticEnv()
+	if b.vipr != nil && b.logger != nil {
+		err := SetCmdServerContext(cmd, b.vipr, b.logger)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return cmd, nil
+}
 
-	return rootCmd.Execute()
+func (b *RootCmdBuilder) Build(
```

Comment:
Shall we add build options in a follow up? To allow to set a custom logger easily instead of ours

Reply:
Looks done with the options now 👌🏾